File: resources/public/js/app/melins.js
(function ($, signs) {
    var coords = signs.coords;

    $(document).ready(function () {
        $("#text-input").keyup(function () {
            var canvas = $("#canvas")[0];
            var text = $(this).val();
            var words = textToWords(text);
            var wordSet = words.map(signSet);


            var dimensions = {
                margin: {x: 60, y: 120},
                border: {x: canvas.width, y: canvas.y}
            };

            var placedWords = placeWords(wordSet, dimensions);

            draw(canvas, placedWords);
        });
    });

    var draw = function (canvas, words) {
        var context = canvas.getContext("2d");
        context.clearRect(0, 0, canvas.height * 3, canvas.width * 3);

        words.forEach(function (word) {
            word.signs.forEach(function (signSet) {
                drawSign(context, signSet.sign, signSet.x + word.x, signSet.y + word.y);
            });
        });
    };

    var drawSign = function (context, sign, x, y) {
        var img = new Image();
        img.src = "img/" + sign + ".png";
        img.onload = function () {
            context.drawImage(this, x, y);
        };
    };

    var textToWords = function (text) {
        return text.trim().split(" ").map(function (word) {
            return word.split("").filter(function (sign) {
                return sign in coords;
            });
        });
    };

    var signSet = function (word) {
        var x = 0,
            y = 0;

        var signs = word.map(function (sign) {
            x += coords[sign].end.x - coords[sign].start.x;
            y += coords[sign].end.y - coords[sign].start.y;
            return {
                sign: sign,
                x: x - coords[sign].end.x,
                y: y - coords[sign].end.y
            };
        });

        return {
            x: 0,
            y: 0,
            signs: signs,
            width: signs.length === 0 ? 0 : Math.abs(signs[signs.length - 1].x - signs[0].x)
        };
    };

    var placeWords = function (words, dim) {
        var marginX = 60;
        var marginY = 120;
        var x = marginX;
        var y = marginY;
        var placedWords = [];

        words.forEach(function (word) {
            if (word.width + x > dim.border.x - dim.margin.x && x != dim.margin.x) {
                y += 60;
                x = marginX;
            }
            placedWords.push({
                x: x,
                y: y,
                signs: word.signs
            });
            x += word.width + 30;
        });

        return placedWords;
    };

})(jQuery, signs);

File: src/melins_system/app/melins.clj
(ns melins-system.app.melins
  (:require [clojure.string :refer [split]]
            [jayq.core :refer [$] :as jq]))

(defn nil-or [value default]
  (if (nil? value) default value))

(defn tokenize [word]
  (filter coords (map str word)))

(defn ->words [text]
  (map tokenize (split (.trim text) #" +")))

(defn word-width [word]
  (if (empty? word)
    0
    (Math/abs (- (:x (last word))
                 (:x (first word))))))

(defn sign-map [word]
  (loop [mapped [] tokens word x 0 y 0]
    (if (empty? tokens)
      mapped
      (let [token (first tokens)
            {{start-x :x start-y :y} :start {end-x :x end-y :y} :end} (coords token)
            sign-map {:sign token
                      :x (+ x start-x)
                      :y (+ y start-y)}]
        (recur (conj mapped sign-map)
               (rest tokens)
               (- (+ x end-x) start-x)
               (- (+ y end-y) start-y))))))

(defn word-map [word]
  (let [signs (sign-map word)]
    {:x 0
     :y 0
     :signs signs
     :width (word-width signs)}))

(defn word-set [text]
  (let [words (->words text)
        word-maps (mapv word-map words)]
    word-maps))

(ns melins.core
  (:require [jayq.core :refer [$] :as jq]
            [melins-system.handler :refer [word-set]]))

(defn draw-sign [context sign]
  (let [img (Image.)
        uri (str "img/" (:sign sign) ".png")]
    (set! (.-src img) uri)
    (set! (.-onload img) (fn [] (.drawImage context img (:x sign) (:y sign))))))

(defn draw-word [context word]
  (map #(draw-sign context %) word))

(defn draw [canvas words]
  (let [context (.getContext canvas "2d")]
    (.clearRext context 0 0 (* 3 (.height canvas)) (* 3 (.height canvas)))
    (dorun (map #(draw-word context %) words))))

(defn render [canvas text]
  (draw canvas (word-set text)))

(defn init []
  (jq/on ($ :#text-input)
         [:keyup]
         #(render ($ "#canvas")
                  (jq/val ($ :#text-input)))))

(jq/document-ready init)
